File: scrimage-core/src/test/scala/com/sksamuel/scrimage/ScalingDeadlockTest.scala
package com.sksamuel.scrimage

import org.scalatest.{Matchers, WordSpec}
import scala.concurrent.{Await, Future}
import scala.concurrent.duration._

class ScalingDeadlockTest extends WordSpec with Matchers {

  import scala.concurrent.ExecutionContext.Implicits.global

  val in: Image = Image.fromStream(getClass.getResourceAsStream("/com/sksamuel/scrimage/bird.jpg"))

  "image scale" should {
    "not deadlock on multiple concurrent scales" in {
      val futures = for ( _ <- 0 until 50 ) yield {
        Future {
          in.scaleTo(200, 200)
        }
      }
      val images = Await.result(Future.sequence(futures), 1.minute)
      images.size shouldBe 50
    }
  }
}

File: scrimage-core/src/test/kotlin/com/sksamuel/scrimage/core/ScalingDeadlockTest.kt
package com.sksamuel.scrimage.core

import com.sksamuel.scrimage.ImmutableImage
import io.kotest.core.spec.style.FunSpec
import io.kotest.inspectors.forAll
import io.kotest.matchers.shouldBe
import java.util.concurrent.CountDownLatch
import kotlin.concurrent.thread

class ScalingDeadlockTest : FunSpec({

   val image = ImmutableImage.fromStream(javaClass.getResourceAsStream("/com/sksamuel/scrimage/bird.jpg"))

   test("image scale should not deadlock on multiple concurrent scales") {
      val images = mutableListOf<ImmutableImage>()
      val latch = CountDownLatch(50)
      repeat(50) {
         thread {
            images.add(image.scaleTo(200, 200))
            latch.countDown()
         }
      }
      latch.await()
      images.size shouldBe 50
      images.forAll { it.width shouldBe 200 }
   }

})
